### DISCUSSION TITLE
Problems with camera, owl is not working

### DISCUSSION THREAD

**Discussion by cyberialyr777** [geezacoleman/OpenWeedLocator#117] [Q&A]
So I'm building owl for an school project. I've already built it and I'm currently in the installation of software. I have my virtual environment and I have followed instructions neatly. The only difference is that my user is mafer and not pi so I've changed paths like in owl_boot.sh.  
![imagen_2024-01-22_200842997](https://github.com/geezacoleman/OpenWeedLocator/assets/112129254/7dafde80-66ab-48c2-a7ca-2f559f9ee714)
and in the beggining of owl.py
![imagen_2024-01-22_201003050](https://github.com/geezacoleman/OpenWeedLocator/assets/112129254/1ac69975-c70b-4ee4-9724-933ed590ce2e)
Do i need to change my paths somewhere else?

Now I want to focus the camera but when I execute the command a small window appears and disappears and this errors come up.
<img width="583" alt="image" src="https://github.com/geezacoleman/OpenWeedLocator/assets/112129254/a7363c60-c379-48dc-abd0-1f835f06cbd0">
I've tried to fix the errors by downlading the library but it doesn't make a difference. 
![imagen_2024-01-22_201645874](https://github.com/geezacoleman/OpenWeedLocator/assets/112129254/115163e7-570c-4633-aa5a-d38d04cf9fdc)

I think it is important to mention that I am accessing the raspberry pi through SSH in a remote connection in my laptop because I don't have a monitor or an spare keyborad. But tomorrow in my university I will try to execute it in a monitor, hoping that this is the issue.

Here is a picture of my owl archive.
![imagen_2024-01-22_202026029](https://github.com/geezacoleman/OpenWeedLocator/assets/112129254/eb81318c-ac97-4551-96a3-783b3d782e80)

I'm currently building owl with just one relay and selenoid, maybe this is the problem? (this is a picture early on building, now it is finished with all needed connections, this picture is for illustration)
![WhatsApp Image 2024-01-10 at 23 03 06_3b1de561](https://github.com/geezacoleman/OpenWeedLocator/assets/112129254/3cfdf022-88e6-44bc-9ae9-5fed94f99673)

Also the camera that I am using is a **Raspberry Pi Module 3** and I don't know if that affects. I've been trying to make it work for a week and I don't know what else to do. I would really appreciate some help and don't hesitate in asking me questions please. Thanks in advance and sorry if you don't understand something, English is not my first language.

**Comment by geezacoleman** [comment]
Thanks for building an OWL as a project! Really cool to see it being used

What was your process for software installation? Did you use the original [OWL.img disk image](https://github.com/geezacoleman/OpenWeedLocator?tab=readme-ov-file#version-history) or did you start from scratch? The latest versions of Raspbian do not support the `picamera` library unfortunately, they've shifted to the libcamera-based `picamera2`. The OWL will be transitioned over to that in the next few months, but it currently won't work out the box. We explain a bit more about that [here](https://github.com/geezacoleman/OpenWeedLocator?tab=readme-ov-file#raspian-operating-system-support).

If you want to use the repository as is and still build from scratch, I would install Raspbian Buster instead. This is the latest Raspbian OS that works with picamera and the owl software. But I imagine you might run into some software versioning issues.

Otherwise, the fastest method would be to use the existing disk image and run the `update_owl.sh` script. But if you're using the existing disk image, then the new user name might be causing other problems.

The single relay won't be a problem, but I'm not too sure about opening video feed windows to focus via SSH. See how it goes with the physical monitor. 

Keep the questions coming if you still have issues - happy to help.

**Reply by cyberialyr777** [reply]
Hi thank you for answering :) . I started the installation from scratch. Do yoy know where can I find the Raspbian Buster? I can't find it in the official page of raspberry pi. And do you think that maybe using Raspberry Pi Camera module 3 is a problem? I read somewhere that this camera doesn't work with picamera library. I will try it with the monitor and tell you how it goes. thanks a lot.

**Reply by cyberialyr777** [reply]
I found this page but I'm not sure which SO to download.  or maybe i need o find somewhere else? 
<img width="551" alt="image" src="https://github.com/geezacoleman/OpenWeedLocator/assets/112129254/613c71b3-4a18-49d2-8361-3d1193ec5953">

**Reply by geezacoleman** [reply]
Ah ok - so I imagine it's a much more recent Raspian version. That will be your biggest issue currently as they don't support picamera. OWL will only work with picamera and unfortunately not picamera2.

There are a few options:
1. start again with the owl disk image provided above.
2. use the second link provided above to download Buster from the official [Raspberry Pi archive](https://downloads.raspberrypi.org/raspios_oldstable_armhf/images/raspios_oldstable_armhf-2022-04-07/).
3. change the code in the current `owl.py` version ([Lines 163 - 182](https://github.com/geezacoleman/OpenWeedLocator/blob/main/owl.py#L163-L182)) where the video is started to use picamera2. This isn't too tricky, it's fairly similar to the below:

```python
from picamera2 import Picamera2 # remove the import statement for imutils.video on L10, otherwise it will cause issues.


# add the code below to replace what is between lines 163 - 182
self.size = resolution 
self.framerate = framerate

self.config = {
"format": 'XRGB8888',
"size": self.size
}

# Initialize the camera and stream
self.cam = Picamera2()
self.cam.set_logging(Picamera2.INFO)
self.cam.configure(self.picam2.create_preview_configuration( main=self.config))

self.cam.start()

```
Then change Line 281 from self.cam.read() to self.cam.capture_array()

This should hopefully get it working at least, but it is unoptimised so will probably be slower, and is largely untested, so you'll probably face some issues. In my opinion, probably worth trying the python code change - I think you'd get it working without too much trouble. The other options of downloading and reimaging the SD card will be quite a lengthy processes!

**Reply by cyberialyr777** [reply]
Sorry, I got a little confused. So, you are giving me two options. The first one is to start again with the owl disk image provided above and hopefully it works. And the second one is to download buster from the link you gave me here,
<img width="177" alt="image" src="https://github.com/geezacoleman/OpenWeedLocator/assets/112129254/b03bf6db-3444-453c-8924-69641b5c4dcf">

do the detailed configuration, and change the code in owl.py as you demonstrate me?

**Reply by geezacoleman** [reply]
Ah sorry for the confusion - first two options are just like you say. Either use the old owl disk image (which has everything pretty well set to go after you run update_owl.sh) or download a new Buster image and set up the owl from scratch as you have done here. These two options try to avoid using picamera2 at all, and just stay with picamera.

The third option is to implement a basic picamera2 version of the OWL. So you would use exactly the same set up you have from the start of this discussion and just change the python code I mentioned above. There are probably other ways to fix this, but these three seem the most obvious to me. 

Third option seems preferable at least at the start, but depends on your use case. I think it will run slightly slower because the camera code isn't threaded.

**Reply by geezacoleman** [reply]
Before you start converting the owl.py file with your current setup, I would first try importing picamera2 in a python environment in the terminal and seeing if that works too. If it does then I would recommend continuing with the changes.

**Reply by cyberialyr777** [reply]
Thank you very much. I will try them all and tell you my results.

**Reply by cyberialyr777** [reply]
So far I've tried to change the code in the set up i have from the start of this discussion and this error appears. 
![WhatsApp Image 2024-01-23 at 12 10 05_3bf668dd](https://github.com/geezacoleman/OpenWeedLocator/assets/112129254/0df374ef-fac3-4baf-932f-e42bc3ed94fc)
I aalready downloaded picamera2 in the virtual env. 
![WhatsApp Image 2024-01-23 at 12 10 03_7cb4c6c4](https://github.com/geezacoleman/OpenWeedLocator/assets/112129254/987776f6-0624-41c7-b85f-ee21f82df19c)
Now I'm going to try with the owl disk image with another sd card.

**Reply by cyberialyr777** [reply]
I tried with the owl disk image but the camera won't function. This errors arise :(
![WhatsApp Image 2024-01-23 at 12 29 24_8f5cc6cf](https://github.com/geezacoleman/OpenWeedLocator/assets/112129254/27fa7c76-7463-468e-80b6-5d6f8c498484)

**Reply by geezacoleman** [reply]
Thanks for giving these a go!

For the attempt where you installed picamera2 - were you able to get that working?

For your second attempt with the owl disk image, I did some more digging on the Camera Module 3 and it looks like it only works on [picamera2](https://raspberrypi.stackexchange.com/questions/142527/raspberry-pi-camera-module-3-works-only-on-libcamera) :( so you'll have to stick with the first option of getting picamera2 working on the pi unfortunately! I had thought the camera modules were fairly safe from the libcamera change, but it apparently not. Sorry about that!

So I think the only option then if you want to use the Camera Module 3 is to persist with getting the picamera2 library working on your system, as you have tried doing above. I'm a bit busy getting a thesis done this week, but I should have time early next week to get you a roughly working example on a separate branch - though definitely worth having a go yourself!

**Reply by cyberialyr777** [reply]
Hello! So to answer this question: For the attempt where you installed picamera2 - were you able to get that working? I couldn't. The reason is because for some reason it doesn't recognize the picamera2 library. I installed it in the virtualenv as I showed you and I import it in the owl.py document as you told me but when I execute the command to focus the camera it appears that there is no module named 'picamera2'. I've been investigating different ways to install it and to import it but i haven't found more because the ones I have used are the correct ones. But i will keep trying and investigating. If i find domething I'll tell you. 

Also, thank you so much for helping me and providing advice even though you are very busy. Hope you finish your thesis and have a great result! Have a great day :)
